cpp/src/parquet/types.h

```diff
@@ -89,11 +90,831 @@ struct LogicalType {
+// Sort order for page and column statistics. Types are associated with sort
+// orders (e.g., UTF8 columns should use UNSIGNED) and column stats are
+// aggregated using a sort order. As of parquet-format version 2.3.1, the
+// order used to aggregate stats is always SIGNED and is not stored in the
+// Parquet file. These stats are discarded for types that need unsigned.
+// See PARQUET-686.
+struct SortOrder {
+  enum type { SIGNED, UNSIGNED, UNKNOWN };
+};
+
+namespace schema {
+
+struct DecimalMetadata {
+  bool isset;
+  int32_t scale;
+  int32_t precision;
+};
+
+}  // namespace schema
+
+#define set_decimal_metadata(m___, i___, p___, s___) \
+  {                                                  \
+    if (m___) {                                      \
+      (m___)->isset = (i___);                        \
+      (m___)->scale = (s___);                        \
+      (m___)->precision = (p___);                    \
+    }                                                \
+  }
+
+#define reset_decimal_metadata(m___) \
+  { set_decimal_metadata(m___, false, -1, -1); }
+
+// using PrimitiveType = parquet::Type;
+// using ConvertedType = parquet::LogicalType;
+// using ConvertedDecimalMetadata = parquet::schema::DecimalMetadata;
+// using LogicalType = parquet::LogicalAnnotation;
+
+class PARQUET_EXPORT LogicalAnnotation {
+ public:
+  enum class Type {
+    Unknown = 0,
+    String = 1,
+    Map,
+    List,
+    Enum,
+    Decimal,
+    Date,
+    Time,
+    Timestamp,
+    Interval,
+    Int,
+    Null,
+    JSON,
+    BSON,
+    UUID,
+    None
+  };
+
+  enum class TimeUnit { Unknown = 0, MilliSeconds = 1, MicroSeconds, NanoSeconds };
+
+#define time_unit_string(u___)                                                      \
+  ((u___) == LogicalAnnotation::TimeUnit::MilliSeconds                              \
+       ? "milliseconds"                                                             \
+       : ((u___) == LogicalAnnotation::TimeUnit::MicroSeconds                       \
+              ? "microseconds"                                                      \
+              : ((u___) == LogicalAnnotation::TimeUnit::NanoSeconds ? "nanoseconds" \
+                                                                    : "unknown")))
+
+  // If possible, return an annotation equivalent to the given legacy converted type
+  // (and decimal metatdata if applicable).
+  static std::shared_ptr<const LogicalAnnotation> FromConvertedType(
+      const LogicalType::type converted_type,
+      const schema::DecimalMetadata converted_decimal_metadata = {false, -1, -1});
+
+  // Return the annotation represented by the Thrift intermediary object
+  static std::shared_ptr<const LogicalAnnotation> FromThrift(
+      const format::LogicalType& thrift_logical_type);
+
+  // Return the requested type of annotation ...
+  static std::shared_ptr<const LogicalAnnotation> UnknownType(void);
```

Comment:
The names for both the enumeration and the methods are tricky as there are several confounding factors, including: (1) using all-caps on the enum values and initial caps on the methods works mostly, except that several types (JSON, UUID) are naturally all caps so their methods want to be all caps as well; (2) there's the 'Null' annotation which can't be made all caps in the enum because of course that confuses the compiler; (3) the authors of the Thrift spec resolved that by naming their NullType field 'UNKNOWN', but that conflicts with the C++ library's convention of using UNKNOWN for "we don't have a value" (as opposed to "we have a value and it's Null"); (4) the converted type enum has an NA value that has no corresponding Thrift LogicalType.

I flipped back and forth several times during development. In this pass, I've adopted your suggestion and changed the enum values to all-caps and shortened the factory methods to be like LogicalAnnotation::Decimal(16, 6).  For the Thrift NullType, I've set the enumeration value to NIL (not UNKNOWN as the Thrift spec does) and used UNKNOWN for invalid/NA annotations.

I'll change things around again if you have other naming preferences.